common/comms/src/inbound_message_service/inbound_message_service.rs

```diff
@@ -0,0 +1,196 @@
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::inbound_message_service::{
+    message_context::MessageContext,
+    message_dispatcher::MessageDispatcher,
+    msg_processing_worker::*,
+};
+use std::{marker::Send, thread};
+use tari_comms::connection::{
+    types::SocketType,
+    zmq::{Context, InprocAddress, ZmqEndpoint},
+    ConnectionError,
+};
+use tari_crypto::keys::PublicKey;
+
+/// The maximum number of processing worker threads that will be created by the InboundMessageService
+const MAX_MSG_PROCESSING_WORKERS: u8 = 8;
+
+pub struct InboundMessageService<PubKey>
+where PubKey: PublicKey
+{
+    context: Context,
+    inbound_address: InprocAddress,
+    dealer_address: InprocAddress,
+    node_identity: PubKey,
+    message_dispatcher: MessageDispatcher<MessageContext<PubKey>>,
+}
+
+impl<PubKey: PublicKey + Send + 'static> InboundMessageService<PubKey> {
+    /// Creates a new InboundMessageService that will fairly deal message received on the inbound address to worker
+    /// threads
+    pub fn new(
+        context: Context,
+        inbound_address: InprocAddress,
+        node_identity: PubKey,
+        message_dispatcher: MessageDispatcher<MessageContext<PubKey>>,
+    ) -> Result<InboundMessageService<PubKey>, ConnectionError>
+    {
+        Ok(InboundMessageService {
+            context: context.clone(),
+            inbound_address: inbound_address.clone(),
+            dealer_address: InprocAddress::random(),
+            node_identity,
+            message_dispatcher,
+        })
+    }
+
+    /// Starts the MsgProcessingWorker threads and the InboundMessageService with Dealer in its own thread
+    pub fn start(self) {
+        thread::spawn(move || {
+            // Start workers
+            for _i in 0..MAX_MSG_PROCESSING_WORKERS {
+                MsgProcessingWorker::new(
+                    self.context.clone(),
+                    self.dealer_address.clone(),
+                    self.node_identity.clone(),
+                    self.message_dispatcher.clone(),
+                )
+                .start();
+            }
+
+            // Start dealer
+            loop {
+                match self.context.socket(SocketType::Router) {
```

Comment:
Most socket connections and binds should be replaced with ```Connection``` once binding is better supported.